src/modules/roc_sndio/target_sox/roc_sndio/sox_backend.cpp

```diff
@@ -252,26 +257,51 @@ ISink* SoxBackend::open_sink(core::IAllocator& allocator,
         return NULL;
     }
 
-    if (!sink->open(driver, output)) {
-        return NULL;
+    if (!trigger_auto_detect) {
+        if (!is_fmt_handler_valid(output, driver, filter_flags)) {
+            return NULL;
+        }
+
+        if (!sink->open(driver, output)) {
+            return NULL;
+        }
+        return sink.release();
     }
 
+    do {
+        if (!driver) {
+            if (!select_defaults(driver, output)) {
+                return NULL;
+            }
+
+            if (!is_fmt_handler_valid(output, driver, filter_flags)) {
+                return NULL;
+            }
+        } else {
+            roc_log(LogInfo,
+                    "failed to open sink=(%s) using driver=(%s), trying next driver",
+                    output, driver);
+            driver = get_fallback_driver(driver);
+            if (!driver) {
+                return NULL;
+            }
+        }
+    } while (!sink->open(driver, output));
```

Comment:
The new algorithm looks good, but to me it's a bit hard to decrypt it from the implementation.

If I understand the implementation correctly, it can be simplified to this (pseudo-code):

```
if (driver != NULL) {
    if (!check_handler(driver, output, ...)) {
        return NULL;
    }
    return try_open(driver, output, ...);
} else {
    for (size_t n = 0; n < SIZE(default_driver_priorities); n++) {
        if (!check_handler(default_driver_priorities[n], "default", ...)) {
            continue;
        }
        sink = try_open(default_driver_priorities[n], "default", ...);
        if (!sink) {
            continue;
        }
        return sink;
    }
    return NULL;
}
```

Note that if driver is NULL, then device/file name is also NULL because driver is decoded from IO URI scheme, and ether scheme is non-empty, or the whole URI is empty. Earlier, before URIs were introduced, it was not always true, and there are artefacts in code from that time.

It seems that select_default_driver(), select_default_device(), select_defaults(), get_fallback_driver() can be removed.

(select_default_device can be removed because all drivers in default_driver_priorities are device drivers, and thus it will always return string "default"; we can just use that string).

Also, I think we can rename default_driver_priorities to just default_drivers and add a comment that they're sorted in probing order. And let's remove "null" driver from the list. If no driver found, it's better to fail than to silently do nothing. I don't remember why I added it.

Reply:
Also, could you please add a check that if the user specified a driver (driver != NULL), but is_driver_hidden() returns true for that driver, then open will fail? It seems we forgot to add it earlier.

Reply:
> The new algorithm looks good, but to me it's a bit hard to decrypt it from the implementation.
> 
> If I understand the implementation correctly, it can be simplified to this (pseudo-code):
> 
> ```
> if (driver != NULL) {
>     if (!check_handler(driver, output, ...)) {
>         return NULL;
>     }
>     return try_open(driver, output, ...);
> } else {
>     for (size_t n = 0; n < SIZE(default_driver_priorities); n++) {
>         if (!check_handler(default_driver_priorities[n], "default", ...)) {
>             continue;
>         }
>         sink = try_open(default_driver_priorities[n], "default", ...);
>         if (!sink) {
>             continue;
>         }
>         return sink;
>     }
>     return NULL;
> }
> ```
> 
> Note that if driver is NULL, then device/file name is also NULL because driver is decoded from IO URI scheme, and ether scheme is non-empty, or the whole URI is empty. Earlier, before URIs were introduced, it was not always true, and there are artefacts in code from that time. -- Is it always true that the file name will be NULL when driver is NULL? Because when we have URIs that point to a filepath like 'file:/home/roc/file.wav' ... the driver name is NULL but the device/file name is the input right?. Sorry if I am misunderstanding you, Just wanted to confirm my belief.

Reply:
Oh, really, you're right!

So...

* If driver is not NULL, output is also not NULL (then use given driver and output).
* If driver is NULL, output may be NULL (then find default driver and default output) or not NULL (then find appropriate driver for given output).

Could you please document these rules in comments to IBackend?

In other words, three cases are possible:

`driver == NULL && output == NULL` => loop through default driver list and use "default" output
`driver == NULL && output != NULL` => just open given driver and output (SoX will select driver automatically)
`driver != NULL && output != NULL` => just open given driver and output 

And condition in the pseudo-code above may be fixed as:

```
if (output != NULL) {
    < just open >
} else {
    roc_panic_if_not(driver == NULL);
    < loop >
}
```

The fourth case, `driver != NULL && output == NULL`, is not possible with current command-line interface, if I'm not mistaken. It would mean "use given driver and select default device for it". For device drivers, same can be achieved by using notation `pulse://default`, `alsa://default`, etc. For file drivers, default output makes little sense. Using stdin/stdout ("-") as default input/output doesn't look like a good idea.

So I think it's right that the fourth case is not supported. And since command-line forbids it, a panic should be okay.

Reply:
Okay great, I will do that. Is it okay to add these rules as a multiline comment in IBackend before the declarations of open_sink and open_source?

Reply:
You can add it to class-level comment. IIRC comment should be attached either to a class or a method to be handled by doxygen.